synapse/lib/hiveauth.py

```diff
@@ -1262,6 +1284,114 @@ async def tryPasswd(self, passwd, nexs=True):
+        pattern = complexity.get('special:regex')
+        fullpat.append(pattern)
+
+        match = regex.findall(pattern, passwd, flags=regex.V1)
+        if count > 0 and not match:
+            failures.append(f'Password must contain at least {count} special characters.')
+        elif len(match) < count:
+            failures.append(f'Password must contain at least {count} special characters, {len(match)} found.')
+
+        # Check numbers
+        count = complexity.get('number:count')
+        pattern = complexity.get('number:regex')
+        fullpat.append(pattern)
+
+        match = regex.findall(pattern, passwd, flags=regex.V1)
+        if count > 0 and not match:
+            failures.append(f'Password must contain at least {count} number characters.')
+        elif len(match) < count:
+            failures.append(f'Password must contain at least {count} number characters, {len(match)} found.')
+
+        fullpat = '|'.join(fullpat)
+        full = regex.findall(fullpat, passwd, flags=regex.V1)
+        if (invalid := set(passwd) - set(full)):
+            failures.append(f'Password contains invalid characters: {sorted(list(invalid))}')
+
+        # Check sequences
+        seqlen = complexity.get('sequences')
+        if seqlen > 1:
+            # Convert each character to it's ordinal value so we can look for
+            # forward and reverse sequences in windows of seqlen. Doing it this
+            # way allows us to easily check unicode sequences too.
+            passb = [ord(k) for k in passwd]
+            for offs in range(len(passwd) - (seqlen - 1)):
+                curv = passb[offs]
+                fseq = list(range(curv, curv + seqlen))
+                rseq = list(range(curv, curv - seqlen, -1))
+                window = passb[offs:offs + seqlen]
+                if window == fseq or window == rseq:
+                    failures = f'Password must not contain forward/reverse sequences longer than {seqlen} characters.'
+                    break
+
+        # Check for previous password reuse
+        prevvalu = self.auth.policy.get('previous')
+        if prevvalu:
+            previous = self.info.get('policy:previous', ())
+            for prevshad in previous:
+                if await s_passwd.checkShadowV2(passwd, prevshad):
+                    failures.append(f'Password cannot be the same as previous {prevvalu} password(s).')
+                    break
+
+        if failures:
+            mesg = ['Cannot change password due to the following policy violations:']
+            mesg.extend(f'  - {msg}' for msg in failures)
+            raise s_exc.BadArg(mesg='\n'.join(mesg), failures=failures)
+
+        if prevvalu:
+            # Looks like this password is good, add it to the list of previous passwords
+            previous = self.info.get('policy:previous', ())
+            previous = (shadow,) + previous
+            await self.auth.setUserInfo(self.iden, 'policy:previous', previous[:prevvalu])
```

Comment:
This may be a nexs transaction coming in from a function called with `nexs=False`

Reply:
I added a comment in the PR about that. I didn't do that logic because this whole file is changing anyway and I'm going to have to update for the new logic once the hive PR is merged.